File: Matlab/NeuralNetDriver.m
function [ theta, missclassifications, validationTheta, validationErrors] = NeuralNetDriver(nil, maxNetworkIterations, theta, numTrainingSamples, numValidationSamples, numTestingSamples, improveTheta)
%NEURALNETDRIVER calls TrainNeuralNetwork and evaluates the resulting
%parameters on a testing set. The MNIST handwritten digit dataset will be
%used for training, validation, and testing.
%INPUT: See TRAINNEURALNet
%nil, maxNetworkIterations, theta, numTrainingSamples, numValidationSamples, numTestingSamples, improveTheta
%OUTPUT:
%[theta, missclassifications, validationTheta, validationErrors]

%Generate data and labels from MNIST
[trainData, trainLabels, validationData, validationLabels, testData, testLabels] = GenNeuralNetData('MNIST', numTrainingSamples, numValidationSamples, numTestingSamples);

if isnan(theta) || improveTheta
    %Train a neural network and return theta (the 3D weight matrix)
    [validationTheta, theta, validationErrors] = TrainNeuralNet(trainData, trainLabels, validationData, validationLabels, nil, maxNetworkIterations, theta, improveTheta);
end

%Calculate accuracy on the testing set
[dataWidth, dataHeight, numSamples]= size(testData);
L=length(nil);

%Sigmoid
h=@(x) 1./(1+exp(-x));

%Number of sampled incorrectly classified
missclassifications=0;

for m=1:numSamples
    %Forward Propagation. No need to store activation matrix.
    tempActivationVec = reshape(testData(:, :, m), 1, dataHeight.*dataWidth); %Unroll data;
    for layer = 2:L
        tempActivationVec =  h(tempActivationVec*validationTheta(1:nil(layer-1), 1:nil(layer), layer-1));
    end
    
    %Backpropagation for first layer only (find errors)
    missclassifications = missclassifications + sum(unique(abs(tempActivationVec - testLabels(m, 1:nil(L))) >= .5));
end

end

File: Python/NeuralNetDriver.py
import numpy as np
import time
import GenNeuralNetData
import TrainNeuralNet
import NeuralNetSupportFunctions as ANNSupport

#Calls the training routine and checks the learned network parameters on the test set
def RunNeuralNetDriver(nil, maxNetworkIterations, theta, numTrain, numValidation, numTest, improveTheta):


    #See if a pretrained weight set is given
    if improveTheta:
        if theta==0:
            print("!!!ERROR: Theta is set to zero, but improveTheta is set to one. Ser improveTheta to 0 or load a valid theta weight matrix")

    #Train a network, if applicable
    if theta==0 or improveTheta:
        #Load data from MNIST
        print(">>>Loading Data From MNIST")
        trainData, validationData, testData, trainLabels, validationLabels, testLabels = GenNeuralNetData.loadMNIST(numTrain, numValidation, numTest)
        print(">>>Data Loaded")
        print(">>>Training Neural Network")
        timer=time.time()
        theta, validationErrors = TrainNeuralNet.RunTrainNeuralNet(trainData, trainLabels, validationData, validationLabels, nil, maxNetworkIterations, theta, improveTheta);
        timer=time.time()-timer
        print(">>>Training Complete. Train Time: "+str(timer) + " Seconds")
    print(">>>Testing Neural Network Accuracy")
    timer=time.time()
    #Calculate accuracy on test set
    testShape=testData.shape
    testSamples=testShape[0]
    testPixels=testShape[1]
    L=nil.size #Number of layers in FFANN

    #Initialize misclassifications
    misclassifications=0

    #Calculate predictions
    testErr=0
    for m in np.arange(testSamples):
        #Forward propagation
            activations = ANNSupport.forwardProp(testData[m, 0:testPixels], testPixels, nil, L, theta)
            #Backpropagation for final layer to calculate error
            isIncorrect=ANNSupport.checkPrediction(activations[-1, 0:nil[-1]], testLabels[m], nil, 1)
            misclassifications+=isIncorrect

    percentAccuracy = (testSamples-misclassifications)/testSamples

    timer=time.time()-timer
    print(">>>Testing Complete. Test Time: "+ str(timer) + " Seconds. Misclassified "+str(misclassifications) + " Samples. Dataset Size: " + str(testSamples) + " Samples. Test Accuracy: " + str(100*percentAccuracy) + " Percent.")
    return theta, validationErrors, percentAccuracy

